app/aws/s3.py

```diff
@@ -19,6 +19,70 @@
 JOBS_CACHE_MISSES = "JOBS_CACHE_MISSES"
 
 
+def list_s3_objects():
+    bucket_name = current_app.config["CSV_UPLOAD_BUCKET"]["bucket"]
+    access_key = current_app.config["CSV_UPLOAD_BUCKET"]["access_key_id"]
+    secret_key = current_app.config["CSV_UPLOAD_BUCKET"]["secret_access_key"]
+    region = current_app.config["CSV_UPLOAD_BUCKET"]["region"]
+    session = Session(
+        aws_access_key_id=access_key,
+        aws_secret_access_key=secret_key,
+        region_name=region,
+    )
+    s3 = session.client("s3")
+
+    objects = []
+    try:
+        response = s3.list_objects_v2(Bucket=bucket_name)
+        while True:
+            for obj in response.get("Contents", []):
+                objects.append(obj["Key"])
+            if "NextContinuationToken" in response:
+                response = s3.list_objects_v2(
+                    Bucket=bucket_name,
+                    ContinuationToken=response["NextContinuationToken"],
+                )
+            else:
+                break
+    except Exception as e:
+        current_app.logger.error(
+            f"An error occurred while regenerating cache #notify-admin-1200 {e}"
+        )
+    return objects
+
```

Comment:
Does it make more sense for this to be a generator than for it to make a list of all of the objects and returning them?

It *appears* that this is used only in a `for` loop. Am I incorrect?

Reply:
I'm not sure I understand the question.

We are getting a list of all the S3 object names/locations.   The for loop checks if a given item is already in the cache, and if it is not (and if it has the "phone number" column) we use the S3 file location to download the missing csv file/job.

What would a generator do or look like in this context?

Reply:
I should add that the thing that is expensive is downloading the csv file.  So we get the list of all the keys in the bucket (very fast operation), then we check in our cache to see if the key is present and then only if the key is not present do we do the expensive thing and download the file.

Reply:
```
def list_s3_objects():
    bucket_name = current_app.config["CSV_UPLOAD_BUCKET"]["bucket"]
    access_key = current_app.config["CSV_UPLOAD_BUCKET"]["access_key_id"]
    secret_key = current_app.config["CSV_UPLOAD_BUCKET"]["secret_access_key"]
    region = current_app.config["CSV_UPLOAD_BUCKET"]["region"]
    session = Session(
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key,
        region_name=region,
    )
    s3 = session.client("s3")

    try:
        response = s3.list_objects_v2(Bucket=bucket_name)
        while True:
            for obj in response.get("Contents", []):
                yield obj["Key"])
            if "NextContinuationToken" in response:
                response = s3.list_objects_v2(
                    Bucket=bucket_name,
                    ContinuationToken=response["NextContinuationToken"],
                )
            else:
                break
    except Exception as e:
        current_app.logger.error(
            f"An error occurred while regenerating cache #notify-admin-1200 {e}"
        )
```

This then doesn't make a list object (the `objects` variable) in memory, and just hands off the invidual elements pulled from S3. It's a little cleaner and more efficient than constructing a list, stuffing the entire S3 stuff into it, and handing it off. Not to mention scalable, because if the number of S3 items being pulled grows over time, this still handles everything as expected without using up memory in the process.

Reply:
ok, done